### ISSUE TITLE
open-id endpoint doesn't return port from behind proxy when running behind proxy 

### ISSUE THREAD

**Issue by tlann** [keycloak/keycloak#13379]
### Describe the bug

When running in istio and behind a proxy,  all of the apps starting up and registering using the open id endpoint the open-id endpoint doesn't return the port in the iss claim.  But outside of the k8s going through the reverse proxy does return the port

### Version

18.0.2

### Expected behavior

The iss claim is returned with the correct port

### Actual behavior

No port is included inside the iss claim when connecting to the open id endpoint when not using the proxy.  

### How to Reproduce?

Run in istio with the istio ingress gateway acting as a reverse proxy.

```
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: config
  namespace: foo
spec:
  hosts:
  - "*"
  gateways:
  - gateway
  http:
  - match:
    - uri:
        exact: /auth
    - uri:
        prefix: /auth/
    route:
    - destination:
        host: keycloak
        port:
          number: 8080
      headers:
        request:
          set:
            X-Forwarded-For: ""
            X-Forwarded-Proto: "https"
            X-Forwarded-Host: "${INGRESS_HOST}:${SECURE_INGRESS_PORT}"

```

The Keycloak config is 
```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: keycloak
  namespace: foo
  labels:
    app: keycloak
    version: v1
spec:
  replicas: 1
  selector:
    matchLabels:
      app: keycloak
  template:
    metadata:
      labels:
        app: keycloak
        version: v1
    spec:
      imagePullSecrets:
      - name: dockercfg
      containers:
      - name: keycloak
        image: quay.io/keycloak/keycloak:18.0.1
        imagePullPolicy: Always
        ports:
        - containerPort: 8080
        env:
        - name: KC_DB_URL_HOST
          valueFrom:
            configMapKeyRef:
              name: keycloak-config
              key: db-addr
        - name: KC_DB_USERNAME
          value: keycloak
        - name: KC_DB_PASSWORD
          value: postgres
        - name: KC_PROXY
          value: passthrough
        - name: KC_CACHE
          value: none
        - name: KC_HOSTNAME_STRICT_HTTPS
          value: "false"
        - name: KC_HTTP_ENABLED
          value: "true"
        - name: KC_HOSTNAME_STRICT_BACKCHANNEL
          value: "false"
        - name: KC_HOSTNAME_STRICT
          value: "false"		
---
apiVersion: v1
kind: Service
metadata:
  name: keycloak
  namespace: foo
spec:
  selector:
    app: keycloak
  ports:
    - name: http
      protocol: TCP
      port: 80
      targetPort: 8080
```

Hitting the open id endpoint outside of k8s gets the port `"issuer":"https://10.200.16.74:32611/auth/realms/foo",`

while going to the endpoint from the keycloak container gets it without the port 
```
sh-4.4$ curl http://keycloak.foo:80/auth/realms/foo/.well-known/openid-configuration
{"issuer":"http://keycloak.foo/auth/realms/foo" ... 
```

**Comment by pedroigor** [comment]
@tlann Could you elaborate on why you need port `80` in URLs?

**Comment by tlann** [comment]
We do not want the issuer the same for all requests.  We just want it to include the port all of the time.  

We used to have the port set to port 8080 for the keycloak backend service and that would be returned in the legacy KC.  But now it returns the iss claim without the port.  The default iss checker in Spring checks the port.  We made it port 80 for a temporary workaround because it wasn't returning the port in the iss claim for all of the backend services.

**Comment by pedroigor** [comment]
I see. Could you please add here the console output when the server is starting? I would like to see the message that is printed out related to the hostname configuration. Thanks.

**Comment by tlann** [comment]
Thank you.  Please see below @pedroigor 
```
2022-08-08 16:42:44,159 INFO  [org.keycloak.quarkus.runtime.hostname.DefaultHostnameProvider] (main) Hostname settings: FrontEnd: <request>, Strict HTTPS: false, Path: <request>, Strict BackChannel: false, Admin: <request>, Port: -1, Proxied: true
2022-08-08 16:43:01,761 WARN  [org.infinispan.PERSISTENCE] (keycloak-cache-init) ISPN000554: jboss-marshalling is deprecated and planned for removal
2022-08-08 16:43:02,126 WARN  [org.infinispan.CONFIG] (keycloak-cache-init) ISPN000569: Unable to persist Infinispan internal caches as no global state enabled
2022-08-08 16:43:02,303 INFO  [org.infinispan.CONTAINER] (keycloak-cache-init) ISPN000556: Starting user marshaller 'org.infinispan.jboss.marshalling.core.JBossUserMarshaller'
2022-08-08 16:43:03,740 INFO  [org.infinispan.CONTAINER] (keycloak-cache-init) ISPN000128: Infinispan version: Infinispan 'Triskaidekaphobia' 13.0.9.Final
2022-08-08 16:43:04,470 INFO  [org.infinispan.CLUSTER] (keycloak-cache-init) ISPN000078: Starting JGroups channel `ISPN`
2022-08-08 16:43:04,471 INFO  [org.infinispan.CLUSTER] (keycloak-cache-init) ISPN000088: Unable to use any JGroups configuration mechanisms provided in properties {}. Using default JGroups configuration!
2022-08-08 16:43:05,860 WARN  [org.jgroups.protocols.UDP] (keycloak-cache-init) JGRP000015: the send buffer of socket MulticastSocket was set to 1.00MB, but the OS only allocated 212.99KB
2022-08-08 16:43:05,860 WARN  [org.jgroups.protocols.UDP] (keycloak-cache-init) JGRP000015: the receive buffer of socket MulticastSocket was set to 20.00MB, but the OS only allocated 212.99KB
2022-08-08 16:43:05,860 WARN  [org.jgroups.protocols.UDP] (keycloak-cache-init) JGRP000015: the send buffer of socket MulticastSocket was set to 1.00MB, but the OS only allocated 212.99KB
2022-08-08 16:43:05,861 WARN  [org.jgroups.protocols.UDP] (keycloak-cache-init) JGRP000015: the receive buffer of socket MulticastSocket was set to 25.00MB, but the OS only allocated 212.99KB
2022-08-08 16:43:07,947 INFO  [org.jgroups.protocols.pbcast.GMS] (keycloak-cache-init) keycloak-79dc8cfffc-56tpt-39332: no members discovered after 2009 ms: creating cluster as coordinator
2022-08-08 16:43:08,010 INFO  [org.infinispan.CLUSTER] (keycloak-cache-init) ISPN000094: Received new cluster view for channel ISPN: [keycloak-79dc8cfffc-56tpt-39332|0] (1) [keycloak-79dc8cfffc-56tpt-39332]
2022-08-08 16:43:08,044 INFO  [org.infinispan.CLUSTER] (keycloak-cache-init) ISPN000079: Channel `ISPN` local address is `keycloak-79dc8cfffc-56tpt-39332`, physical addresses are `[172.17.0.10:54220]`
2022-08-08 16:43:11,751 INFO  [org.keycloak.quarkus.runtime.storage.database.liquibase.QuarkusJpaUpdaterProvider] (main) Updating database. Using changelog META-INF/jpa-changelog-master.xml
2022-08-08 16:43:18,892 INFO  [org.keycloak.connections.infinispan.DefaultInfinispanConnectionProviderFactory] (main) Node name: keycloak-79dc8cfffc-56tpt-39332, Site name: null
2022-08-08 16:43:27,135 INFO  [io.quarkus] (main) Keycloak 18.0.1 on JVM (powered by Quarkus 2.7.5.Final) started in 101.590s. Listening on: http://0.0.0.0:8080
2022-08-08 16:43:27,135 INFO  [io.quarkus] (main) Profile prod activated.
2022-08-08 16:43:27,136 INFO  [io.quarkus] (main) Installed features: [agroal, cdi, hibernate-orm, jdbc-h2, jdbc-mariadb, jdbc-mssql, jdbc-mysql, jdbc-oracle, jdbc-postgresql, keycloak, narayana-jta, reactive-routes, resteasy, resteasy-jackson, smallrye-context-propagation, smallrye-health, smallrye-metrics, vault, vertx]


```

**Comment by pedroigor** [comment]
@tlann Did you try setting the `X-Forwarded-Port` header?

If you have `hostname` disabled and `proxy` set you should have the port resolved from that header (if set). So that you have the port added to all URLs in the discovery document.

**Comment by tlann** [comment]
@pedroigor I have not tried that.  The issue is from all of the services within the workload with istio.  External to the cluster,  everything seems to work fine.  
I used the [reverse proxy guide ](https://www.keycloak.org/server/reverseproxy) to setup the gateway into the workload from external.  I didn't see a mention of X-Forwarded-Port in there.  I will try that.  
I took a little time reading what that header is for.  Is there any documentation that describes it in relation to Keycloak?

**Comment by pedroigor** [comment]
@tlann There isn't.  Please, let me know if it works for you so we can close this issue. The `X-Forwarded-Port` header is a "standard" header for passing the port, similar to the ones we mention in docs.

In any case, we are introducing two new options for force a base URL to frontend and admin URLs. See https://github.com/keycloak/keycloak/pull/13681.

By setting `hostname-url` you will be able to force a base URL so that the HTTP scheme, host, and port, are resolved from it. So, no need to force headers.

I think it does not help you much because you don't want a fixed URL though.

**Comment by tlann** [comment]
@pedroigor 
So this issue has to do with all of our services inside a k8s istio service-mesh.   Everything outside of it works dandy because it is going through the ingress which applies the headers and hence returns the correct port.  
However,  my theory is because of the envoy proxy sidecars,  the appropriate headers aren't getting set.  I'm very new to envoy, 
 so this is going to take me some time to understand and implement and I have a lot of other tasks that I'm actually assigned.  
I'm hoping to figure this out next week on the side.

**Comment by tlann** [comment]
We are using a Istio flavored service mesh with Envoy proxies.  I'm not sure of your background, so please excuse me if I'm over explaining things.  
One thing to reference,  is the diagram on [this page](https://istio.io/latest/docs/ops/deployment/architecture/).  All of our containers use the istio proxy(envoy) via the side car pattern.  Where all traffic is proxied through the sidecar.  This has the benefit of making certificate management,  traffic management and lots of other things a lot more simpler.  

We are using an istio-ingress as a proxy for external traffic.  We will more then likely have other things in front of it for production, ,  but you can see the headers being set in my first post above and this serves as a good first step.  The external traffic receives the correct iss with the port as expected.  e.g. When it makes a request to https://k8s-ingresshost:secure-ingress-port/auth  the iss returned is https://k8s-ingresshost:secure-ingress-port/auth 

However,  when our Spring and other services startup and make requests to Keycloak,http://keycloak.foo:80, it doesn't return the port.  To work around this I set the k8s service port to be 80 so that it could be resolved without the port as a workaround.  

When you say setting a unique issuer,  the issue we have is our services startup and get the Keycloak public key as well as get other Keycloak info.  If the iss doesn't match the requested endpoint the detault Spring iss checker throws an exception and the services don't startup.  

If I set the hostname settings via an argument on startup,  the results I saw a month or so ago were that the hostname was always the same when returned and this causes other issues with the frontend clients.  But this was before I had the reverse proxy settings in place at the ingress.

**Comment by pedroigor** [comment]
@tlann I'm not familiar with Istio but I see what you trying to do by running it as a side-car. So, sorry if I'm lost. This part makes sense to me because you have the proxy forcing a specific host and port when forwarding requests to the server.

However, you want to call `http://keycloak.foo:80` (backchannel communication going through the private network, bypassing the proxy) to return the `iss` (and other endpoints) from the discovery document with the `80` port set. However, `80` is the default HTTP port and as such, it is resolved internally as `-1`.

One way you can force `80` to be set in URLs is to set `--hostname-port=80` but you probably don't want that because it would break the first use case where clients go through the proxy and expect the port set there to be set to URLs.

This second use case (internal communication) is what makes me completely lost. I understand you have some constraints and you must set the port in the URLs (even though it resolves to a default port like 80) but I can not think of anything that can help you without breaking the "proxy" use case. Or changing the server to allow forcing a default port like `80` when the request is set using the default HTTP port.

**Comment by tlann** [comment]
@pedroigor That is exactly it.  
So one thing I thought I could try is setup the reverse proxy headers for the envoy sidecar.  The next question becomes do I do it for the Keycloak sidecar or the rest of the service sidecars.  I believe it to be for Keycloak's envoy sidecar.  
I've never done that with Envoy yet and this will take me a little bit of time.  But at the moment,  my priorities have been set to focus on other things.

**Comment by tlann** [comment]
@pedroigor It seems like,  Keycloak is blocking the OpenID endpoint from all of the containers.  Do you think this has anything to do with the proxy mode being set to passthrough? 
I get the following from all of the containers when starting up.  
`Requests GET /auth/realms/foo/.well-known/openid-configuration HTTP/1.1" 403`

**Comment by pedroigor** [comment]
@tlann That endpoint is public so there is no authorization when trying to access it.

**Comment by tlann** [comment]
Thanks.  I found the issue with the 403.  The db image I was using was setup previously by someone else and it was really old.  I updated to our latest and it was meant for a non-service mesh deployment.  I thought I had found everything.  But a diff in the realm export found that the realm configuration parameter require-ssl was set to yes.  This resolved that issue.    Hoping me noting this here helps someone else.
